### ISSUE TITLE
Can't set  Charging Current below 5 amps.

### ISSUE THREAD

**Issue by MrBLJ** [iainbullock/tesla_ble_mqtt_docker#4]
First, thanks for your work this is incredibly useful.

I'm using BLE to control the charging amps on my cars when I have excess solar power. Regularly, I need to use a very low value for amps, such a 1 or even 0. This is possible using the HA integration.

Using the slider provided in MQTT, I can successfully adjust charging amps, and request values below 5. This is reflected in the logs :
![image](https://github.com/iainbullock/tesla_ble_mqtt_docker/assets/12542442/f5bcfb4f-3ffb-4848-ba58-c4e28a9a6399)

However, for each request made for current below 5 amps, the value is not taken into account and 5 amps are used instead.

On my screenshot above, request for 3 and4 amp resulted in the car charging at 5 amps. This this reflected in the app.

**Comment by raphmur** [comment]
@MrBLJ Tesla cannot accept charging amps below 5 amps. You can check it in your Tesla app.

This is not super ideal for low excess solar, I agree. In my regulation I have set a variable which is a "overconsumption threshold", typically 2 amps. And fix my target charge current to be minimum 5 amps. Then if I have

- 2 amps excess solar + threshold = 4 amps, I don't charge
- 3 amps excess solar + threshold = 5 amps, I charge at 5 amps
- 4 amps excess solar + threshold = 6 amps, I charge at 5 amps
- above 5 amps excess solar, the threshold has no effect
- 5 amps excess solar: I charge at 5 amps
- 6 amps excess solar: I charge at 6 amps
- etc.

**Comment by MrBLJ** [comment]
> @MrBLJ Tesla cannot accept charging amps below 5 amps. You can check it in your Tesla app.
> 
> This is not super ideal for low excess solar, I agree. In my regulation I have set a variable which is a "overconsumption threshold", typically 2 amps. And fix my target charge current to be minimum 5 amps. Then if I have
> 
> * 2 amps excess solar + threshold = 4 amps, I don't charge
> * 3 amps excess solar + threshold = 5 amps, I charge at 5 amps
> * 4 amps excess solar + threshold = 6 amps, I charge at 5 amps
> * above 5 amps excess solar, the threshold has no effect
> * 5 amps excess solar: I charge at 5 amps
> * 6 amps excess solar: I charge at 6 amps
> * etc.
> 
> If you send a charge amps request below 5 amps, Tesla will keep the minimum = 5. The other way is also true, eg. I have 32 A max charge, if I send 40A target, the car will register 32A.

Not sure to understand. using the Tesla Integration in home assistant I can charge below 5 amps. My model Y is currently charging at 3 amps and I regularly charge at 1:
![image](https://github.com/iainbullock/tesla_ble_mqtt_docker/assets/12542442/e7e22b13-acf3-4067-86ab-ff8e20387e4f)

Is this a limitation imposed by the BLE protocol ?

**Comment by iainbullock** [comment]
You can definately go below 5 amps (to 1 amp if you wish) with the http proxy method as I used that before the rate limiting started. 

I haven't tested the BLE much as I only made the first release on Monday and I'm now on holiday so I can't test it. 

However I am pretty sure I was able to get down to 1 amp. I might have done it incrementatlly from 5 rather than jumping directly to 1 from a larger current.

Please could someone try doing it incrementally and report back? 

Thanks

**Comment by BogdanDIA** [comment]
The BLE API can go down to 0 Amps. 
0 Amps means no charging but the charger's relay is ON. 
However, I'm not using MQTT but ssh commands from HA to rpi where tesla-control is running.

**Comment by raphmur** [comment]
Hmmm
Ok
I cannot go below in the Tesla app so I assumed it was not possible. With the "old" API, any request below 5 amps would set to 5 amps as effective charging, at least for me.

**Comment by MrBLJ** [comment]
> You can definately go below 5 amps (to 1 amp if you wish) with the http proxy method as I used that before the rate limiting started.
> 
> I haven't tested the BLE much as I only made the first release on Monday and I'm now on holiday so I can't test it.
> 
> However I am pretty sure I was able to get down to 1 amp. I might have done it incrementatlly from 5 rather than jumping directly to 1 from a larger current.
> 
> Please could someone try doing it incrementally and report back?
> 
> Thanks

I tested incrementally, from 1 to 5, each time resetting the amps with the integration from HA. So I did :

### Using HA integration with proxy
Amps Set with the Integration in HA : 1
Amps seen in the app : 1

### Using MQTT
Amps set with MQTT in HA : 1
Container logs :
```
Received MQTT message: tesla_ble/charging-amps &
Set Charging Amps to & requested
Attempt 1/5
Ok
```
Amps seen in the app : 5

▶ Same behavior from 1A to 4A set by MQTT: 5A are set. Works fine above 5.

**Comment by BogdanDIA** [comment]
I just tested the behavior on my system. I am able to go from above 5A to below 5A and the phone app will show the correct Amps after each change. It was always like that with the old API and it is also working as expected in the new API and BLE API.

can it be due to the phone app differences?

**Comment by iainbullock** [comment]
If you check in the car you can also check what the actual current is. I was wondering if different cars have different minimum charging current selectable via BLE. I suppose we need more feedback from other users

**Comment by BogdanDIA** [comment]
These are my later observations as I've done more tests:

I tried with another phone (Android) and I got the same issue reported by @MrBLJ. Going 6A->2A with BLE  **looking at the phone app in the same time**. I could see the current is limited at 5A. Updating the HA will show the 5A and I can see that in the HA logs.

However, doing the same 6A->2A with BLE with the phone app closed, makes the switch to 2A as expected. After that, starting the phone app will show 2A. The app should not be really closed but it should have stayed in the background for a while so it will not be connected to the server.

From this I draw the conclusion(that I suspected for some time) that the limitation is done by the phone app when the server pushes the new current value to the app but the app sends back the 5A to the server so the final is 5A.

------

Note: the first phone I tried with is a very old iPhone 6S that does not show the issue, maybe the app version is different, that is why I haven't seen the 5A limitation. A good phone :)

**Comment by iainbullock** [comment]
That's an excellent observation thanks! I will test further when I get back from holidays. However it seems the phone app has to be closed. I don't mind this but might forget and as you say 5 amps is often too much if using for excess solar control

**Comment by BogdanDIA** [comment]
> That's an excellent observation thanks! I will test further when I get back from holidays. However it seems the phone app has to be closed. I don't mind this but might forget and as you say 5 amps is often too much if using for excess solar control

I mostly use the HA app on the phone so the phone app stays the background. For the last couple of days, I used charging from solar excess with BLE commands and I've seen the current going up and down around the 5A even though the app is in background. So app in background seems to disconnect from the server leaving only the BLE active for authentication. I think we are good. More tests are welcome.

**Comment by MrBLJ** [comment]
I'm sorry I'm afraid I don't understand what you're describing. The _Tesla_ app would be the cause of this behavior, and we should close it in order to set the current below 5 amps ?

Couple questions :
1. How come the HA integration, using the API then, can set the current to 1 amp even with the phone app running ? I've tried adjusting the current to 1A with the integration slider, looking straight at the Tesla app, and it successfully updated.
2. Is this behavior limited to when the phone is connected to the car (bluetooth key marked as connected) ? meaning, if my phone is away, there should not be any adjustment problem.

Observation : With the _Tesla_ app opened on my iPhone, and the charging amps visible in the app, it is _actually_ possible to go below 5 amps via BLE, here's how :

- **First Action** : MQTT slider moved from 6A to 2A.
  **Result** : 
    - Current in the _Tesla_ app and in the HA integration set to 5A. ❌
    - MQTT slider value to 2A. ✅
 
- **Second Action** : MQTT slider moved from 2A to 3A.
  **Result** : 
    - Current in the _Tesla_ app and in the HA integration set to 3A. ✅
    - MQTT slider value to 3A. ✅

There's something fishy here ^^

**Comment by BogdanDIA** [comment]
> I'm sorry I'm afraid I don't understand what you're describing. The _Tesla_ app would be the cause of this behavior, and we should close it in order to set the current below 5 amps ?
> 
> Couple questions :
> 
>     1. How come the HA integration, using the API then, can set the current to 1 amp even with the phone app running ? I've tried adjusting the current to 1A with the integration slider, looking straight at the Tesla app, and it successfully updated.
> 
>     2. Is this behavior limited to when the phone is connected to the car (bluetooth key marked as connected) ? meaning, if my phone is away, there should not be any adjustment problem.
> 
> 
> Observation : With the _Tesla_ app opened on my iPhone, and the charging amps visible in the app, it is _actually_ possible to go below 5 amps via BLE, here's how :
> 
>     * **First Action** : MQTT slider moved from 6A to 2A.
>       **Result** :
>       
>       * Current in the _Tesla_ app and in the HA integration set to 5A. ❌
>       * MQTT slider value to 2A. ✅
> 
>     * **Second Action** : MQTT slider moved from 2A to 3A.
>       **Result** :
>       
>       * Current in the _Tesla_ app and in the HA integration set to 3A. ✅
>       * MQTT slider value to 3A. ✅
> 
> 
> There's something fishy here ^^

1. There are different paths for each of the situation:

a. If you set the Amps from HA it will be like: 
a.1. HA->(FleetAPI over internet)->Tesla Server->(internet)->Car
and from Tesla Server to update the Tesla app:
a.2. Tesla Server->(internet)->Tesla app

b. If you set the Amps with BLE the path is:
b.1 Docker container->(BLE)->Car->(internet)->Tesla Server->**(internet)->Tesla app** <- **in this case the 5A limit occurs due to the app.**
If the Tesla app is not connected to the server, the info will stay on the server and will be provided next time the Tesla app connects which is exactly like in case a.2:
b.2 Tesla Server->(internet)-> Tesla app **<-this does not show the 5A limit**

2. This is related to the internet connection between Tesla app and Tesla server so not related to BLE.

I am not using the MQTT but the underlying function is the same for my method and MQTT method (tesla-control cmd line).

On your "**First action**" I assume the slider shows the desired Amps while the actual Amps are set by the Tesla app as 5A as you can see on the phone. You should check in your car on the tablet the actual Amps when testing this. You are in the case a. above. One important note is that over BLE it is impossible to check the actual value set by a command. 

On the "**Second action**" the 5A limit problem problem does not occur because Tesla app limiting to 5A appears only when crossing the 5A threshold going down (like in your case 6A->2A).

**Comment by BogdanDIA** [comment]
Did some more tests yesterday and I found a good way to overcome the 5A limit when setting the current no matter the state of the app.
In the case of crossing down the threshold like in our case 6A->2A there is need for two commands.
The first one targeting 2A is making the car+server+app actually setting the current to 5A.
The second one targeting 2A is actually setting car+server+app to the required 2A.
Tested it on the car to see the change of the charging current.

So basically the same command needed to be run twice with 2A target.

In the implementation I see in the container the execution exits after first successful attempt so a change that should work is:
tesla_ble/charging-amps)
     echo Set Charging Amps to $msg requested
     send_command "charging-set-amps $msg";; <-**do it twice**
    *)

**Comment by iainbullock** [comment]
Very interesting and many thanks for your detailed investigation and write up. 

As I'm away and can't test it for myself. 

Can you mod the code inside the container and see if it works? Attach using docker exec <container-id> -iT ash and make the changes. The change will persist between restarts of the container 

Thanks

**Comment by BogdanDIA** [comment]
> Very interesting and many thanks for your detailed investigation and write up.
> 
> I'm away and can't test it for myself.
> 
> Can you mod the code inside the container and see if it works? Attach using docker exec -it container-id -ash and make the changes. The change will persist between restarts of the container
> 
> Thanks

As I wrote above, my implementation uses services in HA to run remote commands through ssh so did not have the set-up ready.

Got some time to install docker on rpi and build the ble image. On "docker compose up" got "external volume "tesla_ble_mqtt" not found". Got a fix for it?

Anyhow, I did run the container with docker run and setup the env vars inside it, did fiddle a bit with mqtt and run the scripts directly. I can see now the MQTT integration in HA.

next, I will add my keys inside instead of generating and test.

**Comment by iainbullock** [comment]
Thanks for working on this. The external volume "tesla_ble_mqtt" not found" means you have to create a named volume first. I think you can delete or ignore it and docker will create a volume with a random name instead. I use named volumes as a personal preference maybe I should avoid for public repos

**Comment by MrBLJ** [comment]
> > Very interesting and many thanks for your detailed investigation and write up.
> > I'm away and can't test it for myself.
> > Can you mod the code inside the container and see if it works? Attach using docker exec -it container-id -ash and make the changes. The change will persist between restarts of the container
> > Thanks
> 
> As I wrote above, my implementation uses services in HA to run remote commands through ssh so did not have the set-up ready.
> 
> Got some time to install docker on rpi and build the ble image. On "docker compose up" got "external volume "tesla_ble_mqtt" not found". Got a fix for it?
> 
> Anyhow, I did run the container with docker run and setup the env vars inside it, did fiddle a bit with mqtt and run the scripts directly. I can see now the MQTT integration in HA.
> 
> next, I will add my keys inside instead of generating and test.

You can create the volume with  `docker volume create tesla_ble_mqtt`

**Comment by BogdanDIA** [comment]
Ok, got a bit more time to work.
What I was saying about the volume was that somehow the user should have instructions that a volume needs to be created.

So I added my private key to /data, modified the code in run.sh, rebuilt the image and ran again. 
Below I switched to 6A and then to 1A. Due to the 5sec delay between the attempts I was able to see the update in the Tesla app going to 5A and then to 1A. So it works. Pushing update button in HA also gives the same results.
The implementation does not allow resending same Amps value otherwise you would have seen it working by sending the same command two times.

tesla_ble_mqtt  | Received MQTT message: tesla_ble/charging-amps 6
tesla_ble_mqtt  | Set Charging Amps to 6 requested
tesla_ble_mqtt  | **First Amp set**
tesla_ble_mqtt  | Attempt 1/5
tesla_ble_mqtt  | Ok
tesla_ble_mqtt  | **Second Amp set**
tesla_ble_mqtt  | Attempt 1/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Attempt 2/5
tesla_ble_mqtt  | Ok
tesla_ble_mqtt  | Received MQTT message: tesla_ble/charging-amps 1
tesla_ble_mqtt  | Set Charging Amps to 1 requested
tesla_ble_mqtt  | First Amp set
tesla_ble_mqtt  | Attempt 1/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Attempt 2/5
tesla_ble_mqtt  | Ok
tesla_ble_mqtt  | Second Amp set
tesla_ble_mqtt  | Attempt 1/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Attempt 2/5
tesla_ble_mqtt  | Ok

**Comment by MrBLJ** [comment]
> What I was saying about the volume was that somehow the user should have instructions that a volume needs to be created.

I understand and I agree. I've written a procedure that details the installation. : @iainbullock would it be possible to enable the wiki feature in the repo, or should I create a discussion ?

@BogdanDIA I'm not sure I understand the full scope of what you did, but it seems to be working isn't it ?

**Comment by BogdanDIA** [comment]
It works as expected. There is the diff:

diff --git a/app/run.sh b/app/run.sh
index 915b578..65e1474 100755
--- a/app/run.sh
+++ b/app/run.sh
@@ -75,6 +75,9 @@ do
       
     tesla_ble/charging-amps)
      echo Set Charging Amps to $msg requested
+     echo First Amp set
+     send_command "charging-set-amps $msg"
+     echo Second Amp set
      send_command "charging-set-amps $msg";;
     *)
      echo "Invalid MQTT topic";;

**Comment by raphmur** [comment]
@MrBLJ maybe you can create a pull request with a new "INSTALL.md" file?

**Comment by MrBLJ** [comment]
> @MrBLJ maybe you can create a pull request with a new "INSTALL.md" file?

Done.

**Comment by iainbullock** [comment]
Just landed at my holiday destination! It maybe several hours before I can review this but thanks for your contributions. One question should I reduce the sleep 5 to something less?

**Comment by BogdanDIA** [comment]
> Just landed at my holiday destination! It maybe several hours before I can review this but thanks for your contributions. One question should I reduce the sleep 5 to something less?

I am using 0 seconds as delay and it works the same with 5s. I think the delay should be configurable for cases where there is a really slow machine that runs the BLE. Enjoy your holiday.

**Comment by iainbullock** [comment]
> > Just landed at my holiday destination! It maybe several hours before I can review this but thanks for your contributions. One question should I reduce the sleep 5 to something less?
> 
> I am using 0 seconds as delay and it works the same with 5s. I think the delay should be configurable for cases where there is a really slow machine that runs the BLE. Enjoy your holiday.

I've made a commit to the dev branch to add a $SEND_CMD_RETRY_DELAY option

**Comment by iainbullock** [comment]
> It works as expected. There is the diff:
> 
> ```
> diff --git a/app/run.sh b/app/run.sh
> index 915b578..65e1474 100755
> --- a/app/run.sh
> +++ b/app/run.sh
> @@ -75,6 +75,9 @@ do
>        
>      tesla_ble/charging-amps)
>       echo Set Charging Amps to $msg requested
> +     echo First Amp set
> +     send_command "charging-set-amps $msg"
> +     echo Second Amp set
>       send_command "charging-set-amps $msg";;
>      *)
>       echo "Invalid MQTT topic";;
> ```

Before I make a mod for this, should we only send the command twice if the requested current is less than 5 amps? I want to avoid sending more commands than necessary in case it swamps the bluetooth on some systems

**Comment by raphmur** [comment]
Most likely yes. Otherwise you would overload the BLE sending without any real need.

**Comment by MrBLJ** [comment]
> > > Just landed at my holiday destination! It maybe several hours before I can review this but thanks for your contributions. One question should I reduce the sleep 5 to something less?
> > 
> > 
> > I am using 0 seconds as delay and it works the same with 5s. I think the delay should be configurable for cases where there is a really slow machine that runs the BLE. Enjoy your holiday.
> 
> I've made a commit to the dev branch to add a $SEND_CMD_RETRY_DELAY option

Very nice ! Would switching from the main to the dev branch for the container require to pair a new key ? I’d rather confirm before messing things up on my end ^^

**Comment by iainbullock** [comment]
> > > > Just landed at my holiday destination! It maybe several hours before I can review this but thanks for your contributions. One question should I reduce the sleep 5 to something less?
> 
> > > 
> 
> > > 
> 
> > > I am using 0 seconds as delay and it works the same with 5s. I think the delay should be configurable for cases where there is a really slow machine that runs the BLE. Enjoy your holiday.
> 
> > 
> 
> > I've made a commit to the dev branch to add a $SEND_CMD_RETRY_DELAY option
> 
> 
> 
> Very nice ! Would switching from the main to the dev branch for the container require to pair a new key ? I’d rather confirm before messing things up on my end ^^

No the key is persistent in /data

**Comment by iainbullock** [comment]
If you use dev branch be aware that currently Dockerhub builds are only for linux/amd64 platform. You can make your own builds using Dockerfile. I could change this if you need arm7 for example 

Also things might be broken!

**Comment by iainbullock** [comment]
> If you use dev branch be aware that currently Dockerhub builds are only for linux/amd64 platform. You can make your own builds using Dockerfile. I could change this if you need arm7 for example 
> 
> Also things might be broken!

Easy to change back again though as the key is persistent

**Comment by BogdanDIA** [comment]
> > It works as expected. There is the diff:
> > ```
> > diff --git a/app/run.sh b/app/run.sh
> > index 915b578..65e1474 100755
> > --- a/app/run.sh
> > +++ b/app/run.sh
> > @@ -75,6 +75,9 @@ do
> >        
> >      tesla_ble/charging-amps)
> >       echo Set Charging Amps to $msg requested
> > +     echo First Amp set
> > +     send_command "charging-set-amps $msg"
> > +     echo Second Amp set
> >       send_command "charging-set-amps $msg";;
> >      *)
> >       echo "Invalid MQTT topic";;
> > ```
> 
> Before I make a mod for this, should we only send the command twice if the requested current is less than 5 amps? I want to avoid sending more commands than necessary in case it swamps the bluetooth on some systems

sending two BLE commands is needed from 5A(included) below so it should be ok to send just one command from 6A up.

**Comment by BogdanDIA** [comment]
> > > It works as expected. There is the diff:
> > > ```
> > > diff --git a/app/run.sh b/app/run.sh
> > > index 915b578..65e1474 100755
> > > --- a/app/run.sh
> > > +++ b/app/run.sh
> > > @@ -75,6 +75,9 @@ do
> > >        
> > >      tesla_ble/charging-amps)
> > >       echo Set Charging Amps to $msg requested
> > > +     echo First Amp set
> > > +     send_command "charging-set-amps $msg"
> > > +     echo Second Amp set
> > >       send_command "charging-set-amps $msg";;
> > >      *)
> > >       echo "Invalid MQTT topic";;
> > > ```
> > 
> > 
> > Before I make a mod for this, should we only send the command twice if the requested current is less than 5 amps? I want to avoid sending more commands than necessary in case it swamps the bluetooth on some systems
> 
> sending two BLE commands is needed for target current of 4A(included) down so it should be ok to send just one command from 5A up.

**Comment by BogdanDIA** [comment]
Replying from the phone so some messages are strange.

Continuing, actually two commands are needed only when passing the 5A threshold from up->down. If the actual current is under 5A it is needed only one command. However, it will be practically impossible to know the actual current as HA or Tesla app can change it and the BLE implementation wouldn't know it. So it is better to send two commands all the time the target current is under 5A.

**Comment by iainbullock** [comment]
I've made the change directly to the main branch and published a new release v.1.08. I've taken a small risk that I've broken something as I can't test it whilst I'm away.

If one of you could test the 'latest' tag from Dockerhub I'd appreciate it

**Comment by MrBLJ** [comment]
> I've made the change directly to the main branch and published a new release v.1.08. I've taken a small risk that I've broken something as I can't test it whilst I'm away.
> 
> If one of you could test the 'latest' tag from Dockerhub I'd appreciate it

Containers build fine. ✅
Entities are discovered in HA ✅
Error while sending commands. Logs :
```
tesla_ble_mqtt_docker by Iain Bullock 2024 https://github.com/iainbullock/tesla_ble_mqtt_docker
Inspiration by Raphael Murray https://github.com/raphmur
Instructions by Shankar Kumarasamy https://shankarkumarasamy.blog/2024/01/28/tesla-developer-api-guide-ble-key-pair-auth-and-vehicle-commands-part-3
Configuration Options are:
[redacted by author]
Setting up auto discovery for Home Assistant
Listening to MQTT
Received MQTT message: tesla_ble/charging-amps 1
Set Charging Amps to 1 requested
Attempt 1/5
Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
Attempt 2/5
Error: context deadline exceeded
Attempt 3/5
Error: context deadline exceeded
Attempt 4/5
Error: context deadline exceeded
Attempt 5/5
Error: context deadline exceeded
Received MQTT message: tesla_ble/charging-amps 6
Set Charging Amps to 6 requested
Attempt 1/5
Error: context deadline exceeded
Attempt 2/5
```

The RPI hasn't moved, it's approximately 2m from the car. `Error: context deadline exceeded` did not appear previously.

**Comment by BogdanDIA** [comment]
Is your car awake?

**Comment by MrBLJ** [comment]
> Is your car awake?

Yes, here's what's happening :

![behavior](https://github.com/iainbullock/tesla_ble_mqtt_docker/assets/12542442/85c06f9d-c78b-4abe-95ce-05b30f1af18f)

**Comment by BogdanDIA** [comment]
Seems you are not using the latest changes. You should be able to see two commands together with two messages like:
First Amp set
Second Amps set

**Comment by MrBLJ** [comment]
> Seems you are not using the latest changes. You should be able to see two commands together with two messages like: First Amp set Second Amps set

Correct, looks like `docker container prune` and `docker image prune` are not what's needed. `docker compose pull` then `docker compose up -d` did the trick.

Looks like to be a nice workaround !

![behavior-2](https://github.com/iainbullock/tesla_ble_mqtt_docker/assets/12542442/7d5936a9-6373-4466-82dd-7d8a7212d83f)

**Comment by MrBLJ** [comment]
I think we can close this one. An improvement would be, if and when we can wake the car via BLE, to send a wake command if an error of type `context deadline exceeded` is encountered, which seems to be the case when the car is asleep. That's another discussion though.

**Comment by iainbullock** [comment]
Thanks everyone for your contributions on this one
